File: backend/app/services/proctoring/catalog.py
"""The proctoring event vocabulary and the consequence path of each entry.

THREE PATHS (proctoring-spec-doc.md section 4.0), decided HERE and only here:

    A   immediate termination. Proctoring itself has been defeated, so the
        assessment can no longer be considered valid. No warning, no
        recruiter preference consulted.
    B   the shared three-warning counter. Shared across every Path B type:
        a tab switch, a phone and a second person together use all three,
        deliberately, because separate counters would allow twelve incidents
        before any consequence.
    C   logged only. Never a warning, never a termination.

WHO EMITS WHAT
--------------
The browser emits what only the browser can see: detections, focus, permission
changes, its own integrity. The server derives what only the server can know:
whether an identity mismatch is the second consecutive one, whether a second
voice was heard in two consecutive chunks, whether a heartbeat gap opened,
whether an answer was typed unusually against the candidate's own baseline.
`client_emittable` says which; an event a client sends under a server-derived
type is refused at ingestion rather than trusted.

THREE IDENTIFIERS ARE ADDITIONS TO THE SPECIFICATION'S CATALOG, each for a
rule the specification states in prose without naming an event:

    MONITORING_INTERRUPTED   section 9's heartbeat gap, which "appears in the
                             report"
    INTEGRITY_CHECK_FAILED   section 9's integrity self-check, "failure =
                             event, and after 60 seconds = Path A"
    IDENTITY_CHECK_MISMATCH  one of section 3.3's two consecutive mismatches,
                             recorded so the report can say how the
                             termination was reached
    CAMERA_STREAM_FAILED     section 4.1's unrecovered stream failure

These identifiers are INTERNAL. None of them reaches a recruiter; the report
speaks in `phrasing.py`'s sentences.
"""
from __future__ import annotations

from dataclasses import dataclass

__all__ = [
    "PATH_A",
    "PATH_B",
    "PATH_C",
    "EventSpec",
    "CATALOG",
    "spec_for",
    "is_known",
    "CLIENT_EMITTABLE",
    "TERMINATING",
    "WARNING_EVENTS",
    "LOGGED_ONLY",
    "GROUP_SCREEN",
    "GROUP_CAMERA",
    "GROUP_AUDIO",
    "GROUP_ANSWERS",
    "GROUP_SYSTEM",
    "CLIENT_RATE_LIMIT_PER_MINUTE",
]

PATH_A = "A"
PATH_B = "B"
PATH_C = "C"

#: Report findings groups (section 7.2). Every event belongs to exactly one.
GROUP_SCREEN = "screen_browser"
GROUP_CAMERA = "camera"
GROUP_AUDIO = "audio"
GROUP_ANSWERS = "answer_patterns"
#: Monitoring interruptions and quality notes. Rendered in the summary and
#: the activity log rather than under a findings heading of their own.
GROUP_SYSTEM = "system"


@dataclass(frozen=True)
class EventSpec:
    """One catalog entry.

    `cooldown` names the `ProctoringConfig` field holding the seconds during
    which a repeat of this type (or of its `cooldown_key` family) is logged
    but does not warn again. None means no cooldown applies.
    `once_per_session` means the first occurrence takes the path and every
    later one is recorded on Path C; that is how "more than one display" is
    reported without burning three warnings in three minutes.
    """

    event_type: str
    path: str
    group: str
    client_emittable: bool
    cooldown: str | None = None
    cooldown_key: str | None = None
    once_per_session: bool = False

    @property
    def terminates(self) -> bool:
        return self.path == PATH_A

    @property
    def warns(self) -> bool:
        return self.path == PATH_B


_ENTRIES: tuple[EventSpec, ...] = (
    # ── Path A: immediate termination (section 4.1) ──────────────────────────
    EventSpec("IDENTITY_MISMATCH", PATH_A, GROUP_CAMERA, client_emittable=False),
    EventSpec("CAMERA_OBSTRUCTED", PATH_A, GROUP_CAMERA, client_emittable=True),
    EventSpec("FACE_ABSENT_EXTENDED", PATH_A, GROUP_CAMERA, client_emittable=True),
    EventSpec("CAMERA_PERMISSION_LOST", PATH_A, GROUP_CAMERA, client_emittable=True),
    EventSpec("MIC_PERMISSION_LOST", PATH_A, GROUP_AUDIO, client_emittable=True),
    EventSpec("CAMERA_STREAM_FAILED", PATH_A, GROUP_CAMERA, client_emittable=True),
    EventSpec("INTEGRITY_CHECK_FAILED", PATH_A, GROUP_SYSTEM, client_emittable=True),
    # ── Path B: the shared warning counter (section 4.2) ─────────────────────
    EventSpec("FULLSCREEN_EXITED", PATH_B, GROUP_SCREEN, client_emittable=True),
    EventSpec("WINDOW_FOCUS_LOST", PATH_B, GROUP_SCREEN, client_emittable=True),
    EventSpec(
        "DEVICE_DETECTED_PHONE", PATH_B, GROUP_CAMERA, client_emittable=True,
        cooldown="object_cooldown_seconds", cooldown_key="DEVICE_DETECTED_PHONE",
    ),
    EventSpec(
        "DEVICE_DETECTED_LAPTOP", PATH_B, GROUP_CAMERA, client_emittable=True,
        cooldown="object_cooldown_seconds", cooldown_key="DEVICE_DETECTED_LAPTOP",
    ),
    EventSpec(
        "DEVICE_DETECTED_SCREEN", PATH_B, GROUP_CAMERA, client_emittable=True,
        cooldown="object_cooldown_seconds", cooldown_key="DEVICE_DETECTED_SCREEN",
    ),
    EventSpec(
        "SECOND_PERSON_DETECTED", PATH_B, GROUP_CAMERA, client_emittable=True,
        cooldown="second_person_cooldown_seconds", cooldown_key="SECOND_PERSON_DETECTED",
    ),
    EventSpec("SECOND_VOICE_DETECTED", PATH_B, GROUP_AUDIO, client_emittable=False),
    EventSpec(
        "FACE_ABSENT_MODERATE", PATH_B, GROUP_CAMERA, client_emittable=True,
        cooldown="face_absent_moderate_cooldown_seconds", cooldown_key="FACE_ABSENT_MODERATE",
    ),
    EventSpec(
        "MULTIPLE_DISPLAYS_DETECTED", PATH_B, GROUP_SCREEN, client_emittable=True,
        once_per_session=True,
    ),
    # ── Path C: logged only (section 4.3, 4.4, 9) ────────────────────────────
    EventSpec("FACE_ABSENT_BRIEF", PATH_C, GROUP_CAMERA, client_emittable=True),
    EventSpec("IDENTITY_CHECK_MISMATCH", PATH_C, GROUP_CAMERA, client_emittable=True),
    EventSpec(
        "LOW_LIGHT", PATH_C, GROUP_CAMERA, client_emittable=True,
        cooldown="low_light_cooldown_seconds", cooldown_key="LOW_LIGHT",
    ),
    EventSpec("BLOCKED_ACTION_ATTEMPTED", PATH_C, GROUP_SCREEN, client_emittable=True),
    EventSpec("FAST_TEXT_ENTRY", PATH_C, GROUP_ANSWERS, client_emittable=False),
    EventSpec("UNIFORM_TEXT_ENTRY", PATH_C, GROUP_ANSWERS, client_emittable=False),
    EventSpec("LOW_TYPED_RATIO", PATH_C, GROUP_ANSWERS, client_emittable=False),
    EventSpec("MOUSE_BEHAVIOR_DEVIATION", PATH_C, GROUP_ANSWERS, client_emittable=False),
    EventSpec("AI_TEXT_SIGNAL", PATH_C, GROUP_ANSWERS, client_emittable=False),
    EventSpec("SESSION_QUALITY_DEGRADED", PATH_C, GROUP_SYSTEM, client_emittable=True),
    EventSpec("MONITORING_INTERRUPTED", PATH_C, GROUP_SYSTEM, client_emittable=False),
    EventSpec("INTEGRITY_CHECK_WARNING", PATH_C, GROUP_SYSTEM, client_emittable=True),
    EventSpec("CAMERA_STREAM_INTERRUPTED", PATH_C, GROUP_CAMERA, client_emittable=True),
)

CATALOG: dict[str, EventSpec] = {entry.event_type: entry for entry in _ENTRIES}

CLIENT_EMITTABLE: frozenset[str] = frozenset(
    entry.event_type for entry in _ENTRIES if entry.client_emittable
)
TERMINATING: frozenset[str] = frozenset(e.event_type for e in _ENTRIES if e.terminates)
WARNING_EVENTS: frozenset[str] = frozenset(e.event_type for e in _ENTRIES if e.warns)
LOGGED_ONLY: frozenset[str] = frozenset(e.event_type for e in _ENTRIES if e.path == PATH_C)

#: Abuse ceiling on client-emitted events per session per minute. A browser
#: emitting more than this is broken or hostile; the surplus is refused with
#: a 429 rather than stored. Not a threshold about candidate behaviour, so
#: it lives with the vocabulary rather than in Settings.
CLIENT_RATE_LIMIT_PER_MINUTE = 600


def spec_for(event_type: str) -> EventSpec:
    """The catalog entry, or a KeyError naming the unknown identifier."""
    return CATALOG[event_type]


def is_known(event_type: str) -> bool:
    return event_type in CATALOG


if len(CATALOG) != len(_ENTRIES):  # pragma: no cover - an import-time contract
    raise ImportError("proctoring catalog has a duplicated event type")

File: frontend/lib/proctoring/catalog.ts
/**
 * The events this client may emit, mirrored from
 * `backend/app/services/proctoring/catalog.py` (the entries with
 * `client_emittable=True`) with the consequence path the server assigns to
 * each.
 *
 * THE SERVER DECIDES. The path is carried here for ONE reason: an event on
 * Path A or Path P is flushed the moment it is queued rather than on the next
 * batch tick (`isImmediate`), because the session is about to end or pause and
 * the candidate should learn that from a screen and not from a request that
 * lands seconds later. The client never acts on a path itself; it requests,
 * the server issues.
 *
 * `lib/proctoring/model-assets.test.ts` reads the Python catalog and asserts
 * this table names exactly the client-emittable identifiers, so an event
 * added on one side without the other fails a test rather than a request.
 */

export type ConsequencePath = "A" | "B" | "C" | "P";

export const CLIENT_EVENTS = {
  // Path A: proctoring itself defeated; immediate termination.
  CAMERA_OBSTRUCTED: "A",
  FACE_ABSENT_EXTENDED: "A",
  INTEGRITY_CHECK_FAILED: "A",
  // Path P (Phase 3, 2026-09-24): a lost camera or microphone PAUSES the
  // assessment for the server's grace instead of ending it. These four were
  // Path A until then (the microphone's stream failure had no event at all);
  // a third loss, or no recovery inside the grace, is what ends a session
  // now, and the server derives both.
  CAMERA_PERMISSION_LOST: "P",
  MIC_PERMISSION_LOST: "P",
  CAMERA_STREAM_FAILED: "P",
  MIC_STREAM_FAILED: "P",
  // Path B: the shared three-warning counter.
  FULLSCREEN_EXITED: "B",
  WINDOW_FOCUS_LOST: "B",
  DEVICE_DETECTED_PHONE: "B",
  DEVICE_DETECTED_LAPTOP: "B",
  DEVICE_DETECTED_SCREEN: "B",
  SECOND_PERSON_DETECTED: "B",
  FACE_ABSENT_MODERATE: "B",
  MULTIPLE_DISPLAYS_DETECTED: "B",
  // Path C: logged only.
  FACE_ABSENT_BRIEF: "C",
  IDENTITY_CHECK_MISMATCH: "C",
  LOW_LIGHT: "C",
  BLOCKED_ACTION_ATTEMPTED: "C",
  SESSION_QUALITY_DEGRADED: "C",
  INTEGRITY_CHECK_WARNING: "C",
  // A stream that came back inside `device_glitch_seconds`: logged, and it
  // costs no pause.
  CAMERA_STREAM_INTERRUPTED: "C",
  MIC_STREAM_INTERRUPTED: "C",
  // Every lost device is live again. Logged on Path C; the server reads it to
  // close the open pause (see `device-watch.ts` for when it is sent).
  DEVICE_RECOVERED: "C",
} as const satisfies Record<string, ConsequencePath>;

export type ClientEventType = keyof typeof CLIENT_EVENTS;

/**
 * Whether the queue posts this event the moment it is queued rather than on
 * the next batch tick. Path A because the session is ending; Path P because
 * the assessment is being paused and the candidate should see the grace
 * begin now, not three seconds after their camera went dark; and the recovery
 * because the pause should lift the moment the device is back.
 */
export function isImmediate(eventType: ClientEventType): boolean {
  const path = CLIENT_EVENTS[eventType];
  return path === "A" || path === "P" || eventType === "DEVICE_RECOVERED";
}

/**
 * The COCO classes the specification says to act on (section 3.1) and the
 * event each one requests. `book` is listed there too, but the catalog has no
 * event for it, so it is discarded in the worker with the other classes
 * rather than reported under a type that means something else.
 */
export const OBJECT_EVENT_FOR_LABEL: Readonly<Record<string, ClientEventType>> = {
  "cell phone": "DEVICE_DETECTED_PHONE",
  laptop: "DEVICE_DETECTED_LAPTOP",
  tv: "DEVICE_DETECTED_SCREEN",
};

/** The class whose count, with the face count, decides a second person. */
export const PERSON_LABEL = "person";
